File: apps/mobile/lib/core/contracts/module_apis.dart
import 'trace_models.dart';
import 'trace_result.dart';

final class EnrollRequest {
  const EnrollRequest({
    required this.tag,
    required this.image,
    required this.roi,
  });

  final String tag;
  final ImageInput image;
  final NormalizedRect roi;
}

final class EnrollResponse {
  const EnrollResponse({
    required this.objectId,
    required this.referenceId,
    required this.qualityScore,
    required this.embeddingCount,
    required this.warnings,
  });

  final String objectId;
  final String referenceId;
  final double qualityScore;
  final int embeddingCount;
  final List<String> warnings;
}

abstract interface class EnrollmentApi {
  Future<TraceResult<EnrollResponse>> enroll(EnrollRequest request);
}

abstract interface class VisualEncoder {
  Future<TraceResult<VisualEmbedding>> encode(
    ImageInput image, {
    NormalizedRect? roi,
  });
}

final class RecognizeRequest {
  const RecognizeRequest({
    required this.image,
    required this.references,
    this.minimumSimilarity = 0.75,
    this.maximumResults = 5,
  });

  final ImageInput image;
  final List<ObjectReference> references;
  final double minimumSimilarity;
  final int maximumResults;
}

final class RecognizeResponse {
  const RecognizeResponse({
    required this.detections,
    required this.processingTime,
    required this.modelVersion,
  });

  final List<ObjectDetection> detections;
  final Duration processingTime;
  final String modelVersion;
}

abstract interface class RecognitionApi {
  Future<TraceResult<RecognizeResponse>> recognize(RecognizeRequest request);
}

final class RecordSightingRequest {
  const RecordSightingRequest({
    required this.objectId,
    required this.detectedAt,
    required this.confidence,
    required this.boundingBox,
    required this.location,
    required this.evidenceImage,
  });

  final String objectId;
  final DateTime detectedAt;
  final double confidence;
  final NormalizedRect? boundingBox;
  final GeoFix? location;
  final ImageInput? evidenceImage;
}

final class RecordSightingResponse {
  const RecordSightingResponse({
    required this.sightingId,
    required this.created,
    required this.deduplicatedWith,
  });

  final String sightingId;
  final bool created;
  final String? deduplicatedWith;
}

final class FindLastSeenResponse {
  const FindLastSeenResponse({
    required this.objectId,
    required this.tag,
    required this.lastSeen,
  });

  final String objectId;
  final String tag;
  final Sighting? lastSeen;
}

abstract interface class MemoryApi {
  Future<TraceResult<RecordSightingResponse>> recordSighting(
    RecordSightingRequest request,
  );

  Future<TraceResult<FindLastSeenResponse>> findLastSeen(String objectId);

  Future<TraceResult<List<Sighting>>> getTimeline(
    String objectId, {
    int limit = 50,
  });
}

File: apps/android/core/contracts/src/main/kotlin/com/traceapp/core/contracts/ModuleApis.kt
package com.traceapp.core.contracts

data class EnrollRequest(
    val tag: String,
    val image: ImageInput,
    val roi: NormalizedRect,
)

data class EnrollResponse(
    val objectId: String,
    val referenceId: String,
    val qualityScore: Float,
    val embeddingCount: Int,
    val warnings: List<String>,
)

interface EnrollmentApi {
    suspend fun enroll(request: EnrollRequest): TraceResult<EnrollResponse>
}

interface VisualEncoder {
    /** Loads the on-device model before the first user operation. */
    suspend fun warmUp(): TraceResult<Unit> = TraceResult.Success(Unit)

    suspend fun encode(
        image: ImageInput,
        roi: NormalizedRect? = null,
    ): TraceResult<VisualEmbedding>
}

data class RecognizeRequest(
    val image: ImageInput,
    val references: List<ObjectReference>,
    val minimumSimilarity: Float = 0.75f,
    val maximumResults: Int = 5,
)

data class RecognizeResponse(
    val detections: List<ObjectDetection>,
    val processingTimeMillis: Long,
    val modelVersion: String,
)

interface RecognitionApi {
    suspend fun recognize(request: RecognizeRequest): TraceResult<RecognizeResponse>
}

data class RecordSightingRequest(
    val objectId: String,
    val detectedAtEpochMillis: Long,
    val confidence: Float,
    val boundingBox: NormalizedRect?,
    val location: GeoFix?,
    val evidenceImage: ImageInput?,
)

data class RecordSightingResponse(
    val sightingId: String,
    val created: Boolean,
    val deduplicatedWith: String?,
)

data class FindLastSeenResponse(
    val objectId: String,
    val tag: String,
    val lastSeen: Sighting?,
)

interface MemoryApi {
    suspend fun recordSighting(
        request: RecordSightingRequest,
    ): TraceResult<RecordSightingResponse>

    suspend fun findLastSeen(objectId: String): TraceResult<FindLastSeenResponse>

    suspend fun getTimeline(
        objectId: String,
        limit: Int = 50,
    ): TraceResult<List<Sighting>>
}
